File: src/main/java/com/libs/databinding/BaseDataBindingRecyclerViewAdapter.java
package com.libs.databinding;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * 基于DataBinding的基础RecyclerView适配器
 * Created by wangz on 2018/9/29.
 */

public abstract class BaseDataBindingRecyclerViewAdapter<T> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    public List<T> list = new ArrayList<>();
    public Context mContext;

    public BaseDataBindingRecyclerViewAdapter(Context mContext) {
        this.mContext = mContext;
    }

    public void setDatas(List<T> list) {
        this.list = list;
        notifyDataSetChanged();
    }

    public void addDatas(List<T> list) {
        this.list.addAll(list);
        notifyDataSetChanged();
    }

    public void clearDatas() {
        this.list.clear();
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ViewDataBinding binding = DataBindingUtil.inflate(LayoutInflater.from(mContext), getLayout(), parent, false);
        return new ViewHolderItem(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        ViewDataBinding binding = ((ViewHolderItem) holder).getBinding();
        onBindView(binding, list.get(position), position);
    }

    /**
     * adapter的布局文件
     *
     * @return
     */
    public abstract int getLayout();

    /**
     * 数据绑定 对应 onBindViewHolder中的代码
     *
     * @param dataBinding
     * @param bean
     * @param position
     */
    public abstract void onBindView(ViewDataBinding dataBinding, T bean, int position);

    @Override
    public int getItemCount() {
        return list == null ? 0 : list.size();
    }

    class ViewHolderItem extends RecyclerView.ViewHolder {
        private ViewDataBinding binding;

        public ViewHolderItem(ViewDataBinding headBinding) {
            super(headBinding.getRoot());
            this.binding = headBinding;
        }

        public ViewDataBinding getBinding() {
            return binding;
        }

        public void setBinding(ViewDataBinding binding) {
            this.binding = binding;
        }
    }
}

File: src/main/java/com/libs/databinding/BaseDataBindingRecyclerViewAdapter.kt
package com.libs.databinding

import android.content.Context
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.RecyclerView

import android.view.LayoutInflater
import android.view.ViewGroup

import java.util.ArrayList

/**
 * 基于DataBinding的基础RecyclerView适配器
 *
 * @param <T>        item对应的数据类型
 * @param <TBinding> adapter对应的ViewDataBinding 名称与你的layout.xml命名有关
</TBinding></T> */
abstract class BaseDataBindingRecyclerViewAdapter<T, TBinding : ViewDataBinding>(var context: Context) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    var list: MutableList<T>? = ArrayList()

    /**
     * adapter的布局文件
     *
     * @return
     */
    abstract val layout: Int

    fun setDatas(list: MutableList<T>) {
        this.list = list
        notifyDataSetChanged()
    }

    fun addDatas(list: MutableList<T>) {
        this.list!!.addAll(list)
        notifyDataSetChanged()
    }

    fun clearDatas() {
        this.list!!.clear()
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val binding = DataBindingUtil.inflate<TBinding>(LayoutInflater.from(context), layout, parent, false)
        return ViewHolderItem(binding)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        onBindView((holder as BaseDataBindingRecyclerViewAdapter<*, *>.ViewHolderItem).binding as TBinding, list!![position], position)
    }

    /**
     * 数据绑定 对应 onBindViewHolder中的代码
     *
     * @param binding
     * @param bean
     * @param position
     */
    abstract fun onBindView(binding: TBinding, bean: T, position: Int)

    override fun getItemCount(): Int {
        return if (list == null) 0 else list!!.size
    }

    internal inner class ViewHolderItem(val binding: TBinding) : RecyclerView.ViewHolder(binding.root)
}
